### ISSUE TITLE
Python: c-ares resolver with enabled gevent crashes on Linux

### ISSUE THREAD

**Issue by scarchik** [grpc/grpc#18850]
### What version of gRPC and what language are you using?
gRPC 1.20, Python
 
### What operating system (Linux, Windows,...) and version?
 Linux Ubuntu (Python Docker: 3.7-stretch)
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
Python 3.7
 
### What did you do?
I'm trying to make a call to the service using expiremental gevent functionality:
```
from gevent import monkey
monkey.patch_all()

import grpc.experimental.gevent
grpc.experimental.gevent.init_gevent()

import grpc
import helloworld_pb2
import helloworld_pb2_grpc

if __name__ == '__main__':
    channel = grpc.insecure_channel('hello.domain.com')
    stub = helloworld_pb2_grpc.GreeterStub(channel)
    response = stub.SayHello(helloworld_pb2.HelloRequest(name='you'))
```

### What did you expect to see?
 Response from service.
 
### What did you see instead?
Call to service crashes:
```
D0424 13:55:39.475437200     308 dns_resolver_ares.cc:436]   Using ares dns resolver
I0424 13:55:39.475501500     308 timer_manager.cc:94]        Spawn timer thread
I0424 13:55:39.475555300     308 init.cc:164]                grpc_init(void)
I0424 13:55:39.475651200     313 timer_manager.cc:259]       timers not checked: expect another thread to
I0424 13:55:39.476036000     313 timer_manager.cc:202]       sleep until kicked
I0424 13:55:39.476754800     308 completion_queue.cc:503]    grpc_completion_queue_create_internal(completion_type=0, polling_type=0)
I0424 13:55:39.476831400     308 completion_queue.cc:503]    grpc_completion_queue_create_internal(completion_type=0, polling_type=0)
I0424 13:55:39.477093700     308 ssl_credentials.cc:128]     grpc_ssl_credentials_create(pem_root_certs=(null), pem_key_cert_pair=(nil), verify_options=(nil), reserved=(nil))
I0424 13:55:39.477203900     308 secure_channel_create.cc:198] grpc_secure_channel_create(creds=0x55e705cf5aa0, target=hello.domain.com, args=0x7fc78452c2e0, reserved=(nil))
I0424 13:55:39.478266800     308 client_channel.cc:229]      chand=0x55e705fb6fe8: update: state=IDLE error="No Error" picker=0x55e705fe5a50
I0424 13:55:39.478413600     308 connectivity_state.cc:164]  SET: 0x55e705fb7090 client_channel: IDLE --> IDLE [helper] error=(nil) "No Error"
I0424 13:55:39.478852900     308 client_channel.cc:488]      chand=0x55e705fb6fe8: created resolving_lb_policy=0x55e705c443e0
I0424 13:55:39.479489500     308 channel.cc:285]             grpc_channel_get_target(channel=0x55e705fb6f20)
I0424 13:55:39.479648000     308 credentials.cc:43]          grpc_channel_credentials_release(creds=0x55e705cf5aa0)
I0424 13:55:39.480383100     308 completion_queue.cc:503]    grpc_completion_queue_create_internal(completion_type=0, polling_type=0)
I0424 13:55:39.480459800     308 metadata_array.cc:29]       grpc_metadata_array_init(array=0x7fc784535168)
I0424 13:55:39.480989600     308 metadata_array.cc:29]       grpc_metadata_array_init(array=0x7fc785170ce8)
I0424 13:55:39.481340900     308 call.cc:1910]               grpc_call_start_batch(call=0x55e705fe63e0, ops=0x55e705ea3170, nops=6, tag=0x7fc78452cfc0, reserved=(nil))
I0424 13:55:39.481670300     308 call.cc:1516]               ops[0]: SEND_INITIAL_METADATA(nil)
I0424 13:55:39.481909800     308 call.cc:1516]               ops[1]: SEND_MESSAGE ptr=0x55e705aea420
I0424 13:55:39.482198900     308 call.cc:1516]               ops[2]: SEND_CLOSE_FROM_CLIENT
I0424 13:55:39.482334300     308 call.cc:1516]               ops[3]: RECV_INITIAL_METADATA ptr=0x7fc784535168
I0424 13:55:39.482369500     308 call.cc:1516]               ops[4]: RECV_MESSAGE ptr=0x7fc7845796a0
I0424 13:55:39.482710500     308 call.cc:1516]               ops[5]: RECV_STATUS_ON_CLIENT metadata=0x7fc785170ce8 status=0x7fc785170d00 details=0x7fc785170d08
I0424 13:55:39.483026300     308 call_combiner.cc:116]       ==> grpc_call_combiner_start() [0x55e705fe63f0] closure=0x55e705fe72c8 [executing batch] error="No Error"
I0424 13:55:39.483197400     308 call_combiner.cc:125]         size: 0 -> 1
I0424 13:55:39.483383900     308 call_combiner.cc:134]         EXECUTING IMMEDIATELY
I0424 13:55:39.483639100     308 call.cc:615]                OP[client-channel:0x55e705fe6d50]:  SEND_INITIAL_METADATA{key=3a 70 61 74 68 ':path' value=2f 49 74 65 6d 73 53 65 72 76 69 63 65 2f 67 65 74 5f 61 6c 6c 5f 69 74 65 6d 73 '/HelloService/SayHello'} SEND_MESSA
GE:flags=0x00000000:len=0 SEND_TRAILING_METADATA{} RECV_INITIAL_METADATA RECV_MESSAGE RECV_TRAILING_METADATA
I0424 13:55:39.484109800     308 client_channel.cc:972]      chand=0x55e705fb6fe8 calld=0x55e705fe6d70: adding pending batch at index 0
I0424 13:55:39.484324000     308 client_channel.cc:2814]     chand=0x55e705fb6fe8 calld=0x55e705fe6d70: entering client_channel combiner
I0424 13:55:39.484548300     308 client_channel.cc:2686]     chand=0x55e705fb6fe8 calld=0x55e705fe6d70: LB pick returned QUEUE (connected_subchannel=(nil), error="No Error")
I0424 13:55:39.484737200     308 client_channel.cc:2553]     chand=0x55e705fb6fe8 calld=0x55e705fe6d70: adding to queued picks list
I0424 13:55:39.484910200     308 call_combiner.cc:215]       call_combiner=0x55e705fe63f0: setting notify_on_cancel=0x55e705de9518
I0424 13:55:39.485290900     308 resolving_lb_policy.cc:311] resolving_lb=0x55e705c443e0: starting name resolution
I0424 13:55:39.485491600     308 client_channel.cc:229]      chand=0x55e705fb6fe8: update: state=CONNECTING error="No Error" picker=0x55e705fe5a30
I0424 13:55:39.486124200     308 connectivity_state.cc:164]  SET: 0x55e705fb7090 client_channel: IDLE --> CONNECTING [helper] error=(nil) "No Error"
I0424 13:55:39.486387200     308 client_channel.cc:2686]     chand=0x55e705fb6fe8 calld=0x55e705fe6d70: LB pick returned QUEUE (connected_subchannel=(nil), error="No Error")
D0424 13:55:39.486719500     308 dns_resolver_ares.cc:176]   (c-ares resolver) resolver:0x55e705c918a0 AresDnsResolver::StartLocked() is called.
D0424 13:55:39.486896200     308 grpc_ares_wrapper.cc:536]   (c-ares resolver) request:0x55e705b03950 c-ares grpc_dns_lookup_ares_locked_impl name=hello.domain.com, default_port=https
D0424 13:55:39.487483200     308 grpc_ares_ev_driver.cc:143] (c-ares resolver) request:0x55e705b03950 grpc_ares_ev_driver_create_locked
I0424 13:55:39.487778800     308 socket_utils_common_posix.cc:346] Disabling AF_INET6 sockets because ::1 is not available.
I0424 13:55:39.488282100     308 ev_posix.cc:252]            (fd-trace) fd_create(6, c-ares fd: 6, 0)
Fatal Python error: Segmentation fault

Current thread 0x00007fc78e976400 (most recent call first):
  File "/usr/local/lib/python3.7/site-packages/grpc/_channel.py", line 549 in _blocking
  File "/usr/local/lib/python3.7/site-packages/grpc/_channel.py", line 561 in __call__
  File "wsgi.py", line 54 in <module>
Segmentation fault
```

This code perfectly works on windows with enabled gevent and also work when gevent support is disabled on Linux.

**Comment by gnossen** [comment]
Related:
 - https://github.com/grpc/grpc/pull/18547
 - https://github.com/grpc/grpc/issues/18422

**Comment by lidizheng** [comment]
@scarchik We had a fix in our dev branch #18547. Can you try if our daily build fix your problem?
https://packages.grpc.io/

**Comment by lidizheng** [comment]
The fix is released with v1.20.1 today: https://github.com/grpc/grpc/releases/tag/v1.20.1
@scarchik If your problem isn't fixed by this patch, feel free to reopen this issue :)

**Comment by scarchik** [comment]
It works fine. Thank you!
